File: library/src/main/java/com/alamkanak/weekview/WeekViewTouchHandler.java
package com.alamkanak.weekview;

import android.view.MotionEvent;

import java.util.Calendar;

import static com.alamkanak.weekview.DateUtils.today;
import static java.lang.Math.max;

final class WeekViewTouchHandler {

    private final WeekViewConfigWrapper config;

    WeekViewTouchHandler(WeekViewConfigWrapper config) {
        this.config = config;
    }

    /**
     * Returns the time and date where the user clicked on.
     *
     * @param event The {@link MotionEvent} of the touch event.
     * @return The time and date at the clicked position.
     */
    Calendar getTimeFromPoint(MotionEvent event) {
        final float touchX = event.getX();
        final float touchY = event.getY();

        final float widthPerDay = config.getWidthPerDay();
        final float totalDayWidth = widthPerDay + config.getColumnGap();
        final float originX = config.getCurrentOrigin().x;
        final float timeColumnWidth = config.getTimeColumnWidth();

        final int leftDaysWithGaps = (int) (Math.ceil(originX / totalDayWidth) * (-1));
        float startPixel = originX + totalDayWidth * leftDaysWithGaps + timeColumnWidth;

        final int begin = leftDaysWithGaps + 1;
        final int end = leftDaysWithGaps + config.getNumberOfVisibleDays() + 1;

        for (int dayNumber = begin; dayNumber <= end; dayNumber++) {
            final float start = max(startPixel, timeColumnWidth);

            final boolean isVisibleHorizontally = startPixel + widthPerDay - start > 0;
            final boolean isWithinDay = (touchX > start) & (touchX < startPixel + widthPerDay);

            if (isVisibleHorizontally && isWithinDay) {
                final Calendar day = today();
                day.add(Calendar.DATE, dayNumber - 1);

                final float originY = config.getCurrentOrigin().y;
                final float hourHeight = config.getHourHeight();

                final float pixelsFromZero = touchY - originY - config.getHeaderHeight();
                final int hour = (int) (pixelsFromZero / hourHeight);
                final int minute = (int) (60 * (pixelsFromZero - hour * hourHeight) / hourHeight);
                day.add(Calendar.HOUR, hour + config.getMinHour());
                day.set(Calendar.MINUTE, minute);
                return day;
            }

            startPixel += totalDayWidth;
        }

        return null;
    }

}

File: library/src/main/java/com/alamkanak/weekview/WeekViewTouchHandler.kt
package com.alamkanak.weekview

import android.view.MotionEvent
import com.alamkanak.weekview.date.now
import com.alamkanak.weekview.date.plusDays
import com.alamkanak.weekview.date.withTime
import java.util.*
import kotlin.math.ceil
import kotlin.math.max

internal class WeekViewTouchHandler(
        private val config: WeekViewConfigWrapper
) {

    /**
     * Returns the date and time where the user clicked on.
     *
     * @param event The [MotionEvent] of the touch event.
     * @return The time and date at the clicked position.
     */
    fun getTimeFromPoint(event: MotionEvent): Calendar? {
        val touchX = event.x
        val touchY = event.y

        val widthPerDay = config.widthPerDay
        val totalDayWidth = widthPerDay + config.columnGap
        val originX = config.currentOrigin.x
        val timeColumnWidth = config.timeColumnWidth

        val leftDaysWithGaps = (ceil((originX / totalDayWidth).toDouble()) * -1).toInt()
        var startPixel = originX + totalDayWidth * leftDaysWithGaps + timeColumnWidth

        val begin = leftDaysWithGaps + 1
        val end = leftDaysWithGaps + config.numberOfVisibleDays + 1

        for (dayNumber in begin..end) {
            val start = max(startPixel, timeColumnWidth)

            val isVisibleHorizontally = startPixel + widthPerDay - start > 0
            val isWithinDay = (touchX > start) and (touchX < startPixel + totalDayWidth)

            if (isVisibleHorizontally && isWithinDay) {
                val day = now().plusDays(dayNumber - 1)

                val originY = config.currentOrigin.y
                val hourHeight = config.hourHeight

                val pixelsFromZero = touchY - originY - config.headerHeight
                val hour = (pixelsFromZero / hourHeight).toInt()
                val minutes = (60 * (pixelsFromZero - hour * hourHeight) / hourHeight).toInt()

                return day.withTime(config.minHour + hour, minutes)
            }

            startPixel += totalDayWidth
        }

        return null
    }

}
